## Introduction
---
Hi! Welcome to my first notebook of the 3rd Data Immersion by Alura! In this notebook, i am gonna do the challenges proposed in the lecture and accompany the code done by the instructors! Sometimes I had the curiosity to discover something before happened in the class, so I did early and marked with #personal. My idea is do this as a progression, so as i am doing i will stop the video and write my thoughts. I am not gonna to talk much about code of the instructors, as it was already explained in the video, only mines. Hope you enjoy! :)



In [2]:
import pandas as pd

data_url = 'https://github.com/alura-cursos/imersaodados3/blob/main/dados/dados_experimentos.zip?raw=true'

data = pd.read_csv(data_url, compression= 'zip')

data

id    tratamento  tempo dose      droga     g-0     g-1  \
0      id_000644bb2     com_droga     24   D1  b68db1d53  1.0620  0.5577   
1      id_000779bfc     com_droga     72   D1  df89a8e5a  0.0743  0.4087   
2      id_000a6266a     com_droga     48   D1  18bb41b2c  0.6280  0.5817   
3      id_0015fd391     com_droga     48   D1  8c7f86626 -0.5138 -0.2491   
4      id_001626bd3     com_droga     72   D2  7cbed3131 -0.3254 -0.4009   
...             ...           ...    ...  ...        ...     ...     ...   
23809  id_fffb1ceed     com_droga     24   D2  df1d0a5a1  0.1394 -0.0636   
23810  id_fffb70c0c     com_droga     24   D2  ecf3b6b74 -1.3260  0.3478   
23811  id_fffc1c3f4  com_controle     48   D2  cacb2b860  0.3942  0.3756   
23812  id_fffcb9e7c     com_droga     24   D1  8b87a7a83  0.6660  0.2324   
23813  id_ffffdd77b     com_droga     72   D1  972f41291 -0.8598  1.0240   

          g-2     g-3     g-4  ...    c-90    c-91    c-92    c-93    c-94  \
0     -0.2479 -0.6208 -0.1944  ...  0.2862  0.2584  0.8076  0.5523 -0.1912   
1      0.2991  0.0604  1.0190  ... -0.4265  0.7543  0.4708  0.0230  0.2957   
2      1.5540 -0.0764 -0.0323  ... -0.7250 -0.6297  0.6103  0.0223 -1.3240   
3     -0.2656  0.5288  4.0620  ... -2.0990 -0.6441 -5.6300 -1.3780 -0.8632   
4      0.9700  0.6919  1.4180  ...  0.0042  0.0048  0.6670  1.0690  0.5523   
...       ...     ...     ...  ...     ...     ...     ...     ...     ...   
23809 -0.1112 -0.5080 -0.4713  ...  0.1969  0.0262 -0.8121  0.3434  0.5372   
23810 -0.3743  0.9905 -0.7178  ...  0.4286  0.4426  0.0423 -0.3195 -0.8086   
23811  0.3109 -0.7389  0.5505  ...  0.5409  0.3755  0.7343  0.2807  0.4116   
23812  0.4392  0.2044  0.8531  ... -0.1105  0.4258 -0.2012  0.1506  1.5230   
23813 -0.1361  0.7952 -0.3611  ... -3.3890 -1.7450 -6.6300 -4.0950 -7.3860   

         c-95    c-96    c-97    c-98    c-99  
0      0.6584 -0.3981  0.2139  0.3801  0.4176  
1      0.4899  0.1522  0.1241  0.6077  0.7371  
2     -0.3174 -0.6417 -0.2187 -1.4080  0.6931  
3     -1.2880 -1.6210 -0.8784 -0.3876 -0.8154  
4     -0.3031  0.1094  0.2885 -0.3786  0.7125  
...       ...     ...     ...     ...     ...  
23809 -0.3246  0.0631  0.9171  0.5258  0.4680  
23810 -0.9798 -0.2084 -0.1224 -0.2715  0.3689  
23811  0.6422  0.2256  0.7592  0.6656  0.3808  
23812  0.7101  0.1732  0.7015 -0.6290  0.0740  
23813 -1.4160 -3.5770 -0.4775 -2.1500 -4.2520  

[23814 rows x 877 columns]

In [7]:
#personal

data.tratamento.unique() 

array(['com_droga', 'com_controle'], dtype=object)

In [10]:
#personal

data.droga.unique() 

array(['b68db1d53', 'df89a8e5a', '18bb41b2c', ..., '573c787a2',
       'b2fe3eca7', 'dd4a96d16'], dtype=object)

In [12]:
#personal

len(data.droga.unique())

3289

Early on, when the dataframe was presented, one of my first thoughts was if only really had only two values in the column 'Tratamento', so i went and checked with `unique()`. After a little more video, Vanessa talked about each row being a culture that was being tested with a specific drug. When she said that I looked to the dataset to search for this information, and found it easily in the column 'droga'. After that i procced to investigate hom many drugs we were talking about, so i used `len()` with `unique()` to find how many drugs was in the dataset: 3289 different drugs.

In [15]:
data.head()

id tratamento  tempo dose      droga     g-0     g-1     g-2  \
0  id_000644bb2  com_droga     24   D1  b68db1d53  1.0620  0.5577 -0.2479   
1  id_000779bfc  com_droga     72   D1  df89a8e5a  0.0743  0.4087  0.2991   
2  id_000a6266a  com_droga     48   D1  18bb41b2c  0.6280  0.5817  1.5540   
3  id_0015fd391  com_droga     48   D1  8c7f86626 -0.5138 -0.2491 -0.2656   
4  id_001626bd3  com_droga     72   D2  7cbed3131 -0.3254 -0.4009  0.9700   

      g-3     g-4  ...    c-90    c-91    c-92    c-93    c-94    c-95  \
0 -0.6208 -0.1944  ...  0.2862  0.2584  0.8076  0.5523 -0.1912  0.6584   
1  0.0604  1.0190  ... -0.4265  0.7543  0.4708  0.0230  0.2957  0.4899   
2 -0.0764 -0.0323  ... -0.7250 -0.6297  0.6103  0.0223 -1.3240 -0.3174   
3  0.5288  4.0620  ... -2.0990 -0.6441 -5.6300 -1.3780 -0.8632 -1.2880   
4  0.6919  1.4180  ...  0.0042  0.0048  0.6670  1.0690  0.5523 -0.3031   

     c-96    c-97    c-98    c-99  
0 -0.3981  0.2139  0.3801  0.4176  
1  0.1522  0.1241  0.6077  0.7371  
2 -0.6417 -0.2187 -1.4080  0.6931  
3 -1.6210 -0.8784 -0.3876 -0.8154  
4  0.1094  0.2885 -0.3786  0.7125  

[5 rows x 877 columns]

In [17]:
data.shape

(23814, 877)

In [19]:
data['tratamento']

0           com_droga
1           com_droga
2           com_droga
3           com_droga
4           com_droga
             ...     
23809       com_droga
23810       com_droga
23811    com_controle
23812       com_droga
23813       com_droga
Name: tratamento, Length: 23814, dtype: object

In [20]:
data['tratamento'].unique()

array(['com_droga', 'com_controle'], dtype=object)

Here, the instructor did the same that i thought. That is expected, as exploratory analysis is the first thing that should be done! 😎 

In [27]:
data['tempo'].unique()

array([24, 72, 48], dtype=int64)

In [29]:
data['dose'].unique()

array(['D1', 'D2'], dtype=object)

In [31]:
data['droga'].unique()

array(['b68db1d53', 'df89a8e5a', '18bb41b2c', ..., '573c787a2',
       'b2fe3eca7', 'dd4a96d16'], dtype=object)

In [33]:
data['g-0'].unique()

array([ 1.062 ,  0.0743,  0.628 , ...,  0.3942,  0.666 , -0.8598])

In [38]:
data['g-0'].describe() #pessoal

count    23814.000000
mean         0.248366
std          1.393399
min         -5.513000
25%         -0.473075
50%         -0.008850
75%          0.525700
max         10.000000
Name: g-0, dtype: float64

In [40]:
data['tratamento'].value_counts()

com_droga       21948
com_controle     1866
Name: tratamento, dtype: int64

At this point, 3 challenges were proposed:

## Challenge 01: Why the column 'Tratamento' is so unbalanced?

## Challenge 02: How to show the last 5 entries of a dataframe

## Challenge 03: How to show the proportions of classes in a series, instead of the quantity of values?

In this point, i stopped the video and proceed to answer the challenges. 

The first that I will answer is the **Challenge 2**, because is the easier. To verify the last 5 entries, the Dataframe object has a method called `.tail()`. This method works similarly to the `.head()` method, the difference being that it gets the 5 last rows, if the arguments are left empty.

In [44]:
# Personal 
# Challenge 2 answer

data.tail()

id    tratamento  tempo dose      droga     g-0     g-1  \
23809  id_fffb1ceed     com_droga     24   D2  df1d0a5a1  0.1394 -0.0636   
23810  id_fffb70c0c     com_droga     24   D2  ecf3b6b74 -1.3260  0.3478   
23811  id_fffc1c3f4  com_controle     48   D2  cacb2b860  0.3942  0.3756   
23812  id_fffcb9e7c     com_droga     24   D1  8b87a7a83  0.6660  0.2324   
23813  id_ffffdd77b     com_droga     72   D1  972f41291 -0.8598  1.0240   

          g-2     g-3     g-4  ...    c-90    c-91    c-92    c-93    c-94  \
23809 -0.1112 -0.5080 -0.4713  ...  0.1969  0.0262 -0.8121  0.3434  0.5372   
23810 -0.3743  0.9905 -0.7178  ...  0.4286  0.4426  0.0423 -0.3195 -0.8086   
23811  0.3109 -0.7389  0.5505  ...  0.5409  0.3755  0.7343  0.2807  0.4116   
23812  0.4392  0.2044  0.8531  ... -0.1105  0.4258 -0.2012  0.1506  1.5230   
23813 -0.1361  0.7952 -0.3611  ... -3.3890 -1.7450 -6.6300 -4.0950 -7.3860   

         c-95    c-96    c-97    c-98    c-99  
23809 -0.3246  0.0631  0.9171  0.5258  0.4680  
23810 -0.9798 -0.2084 -0.1224 -0.2715  0.3689  
23811  0.6422  0.2256  0.7592  0.6656  0.3808  
23812  0.7101  0.1732  0.7015 -0.6290  0.0740  
23813 -1.4160 -3.5770 -0.4775 -2.1500 -4.2520  

[5 rows x 877 columns]

My next answers is relative to **Challenge 3**. When i saw that challenge the 'brute way' was the first thing that came to my mind. What i mean by 'brute way'? Well...

In [51]:
proportions = data['tratamento'].value_counts() / data['tratamento'].value_counts().sum()
proportions 

com_droga       0.921643
com_controle    0.078357
Name: tratamento, dtype: float64

Now that i did it, maybe is not so brute at all 🙈 But when i was thinking, i thought that maybe there is a method to do this inside pandas... So i will check the documentation really quick, just a moment 

👨‍💻👨‍💻👩‍💻

ooohhhh, i told you that was a beautiful way! The `.value_counts()` method has a `normalize` parameter! If a set this to `True`, i can get exactly what i want in just ONE function call!!! So beautiful 😍

In [54]:
proportions = data['tratamento'].value_counts(normalize=True)
proportions

com_droga       0.921643
com_controle    0.078357
Name: tratamento, dtype: float64

Now that is 2 answers! Way to go, me! :P And know is the "hard" one, the **Challenge 01**... 

I leave this one last because is the one that i am not totally sure, maybe the instructors will answer latter in the video. In my humble understanding of the the dataset, the reason for the unbalacement is the very nature of the control group. What i mean is, if you are testing, i dont know, 10 drugs for headache and checking they efficiency, you will only need one control sample, as it will be the base to the comparison with the 10 headache drugs. Maybe it isn't, as a said, right now i am not totally sure of it... Well, when i finish, probably  will check my peers works too and the doubt will be cleared! :D

Now that the challenge are answered, lets return to the video 🍿